iotedgedev/deploymentmanifest.py

```diff
@@ -147,3 +186,74 @@ def _get_module_content(self):
             return self.json["moduleContent"]
         else:
             raise KeyError("modulesContent")
+
+    def _validate_deployment_manifest_schema(self):
+        validation_success = True
+        try:
+            self.output.info("Validating schema of deployment manifest.")
+            deployment_schema = json.loads(urlopen(Constants.deployment_manifest_schema_url).read())
+            validator_class = jsonschema.validators.validator_for(deployment_schema)
+            validator = validator_class(deployment_schema)
+            validation_errors = validator.iter_errors(self.json)
+            error_detected = False
+            for error in validation_errors:
+                error_detected = True
+                self.output.info("Deployment manifest schema error: %s. Property path:%s" % (error.message, "->".join(error.path)))
+            if error_detected:
+                self.output.info("Deployment manifest schema validation failed. Please see previous logs for more details")
+            else:
+                self.output.info("Deployment manifest schema validation passed.")
+        except jsonschema.exceptions.SchemaError as schemaErr:
+            self.output.info("Errors found in deployment manifest schema, skip schema validation. Error:%s" % schemaErr.message)
+        except Exception as ex:  # Ignore other non schema validation errors
+            self.output.info("Unexpected error during deployment manifest schema validation, skip schema validation. Error:%s" % ex)
+
+        return validation_success
+
+    # Call _validate_deployment_manifest_schema first. This function assumes createOptions are strings.
+    def _validate_createOptions(self):
+        self.output.info("Start validating createOptions for all modules.")
+        modules = self.get_all_modules()
+        validation_success = True
+        for module_name, module_info in modules.items():
+            try:
+                self.output.info("Validating createOptions for module %s" % module_name)
+                create_options = []
+                if "settings" in module_info and "createOptions" in module_info["settings"]:
+                    create_options_value = module_info["settings"]["createOptions"]
+                    if len(create_options_value) > TWIN_VALUE_MAX_SIZE:
```

Comment:
doesn't check whether is json object before checking len?

Reply:
createOptions may be split so it can be an invalid json string
The function will merge all createOptions and check whether it's a valid json string later: [Line 240](https://github.com/Azure/iotedgedev/blob/chyuan/deployment-manifest-validation/iotedgedev/deploymentmanifest.py#L240)

Reply:
And since the schema defines createOptions as string, the validation relies on the schema to do the check